warehouse/bigquery.go

```diff
@@ -170,6 +168,89 @@ func (bq *BigQuery) LoadToWarehouse(filename string, bundles ...fullstory.Export
 	return bq.waitForJob(job)
 }
 
+// EnsureCorrectExportTable ensures that the all the fields present in the hauser schema are present in the BigQuery table schema
+// If the table exists, it compares the schema of the export table in BigQuery to the schema in hauser and adds any missing fields
+// if the table doesn't exist, it creates a new export table with all the fields specified in hauser
+func (bq *BigQuery) EnsureCorrectExportTable() error {
+	// Get Hauser Schema
+	// this is required if we create a new table or if we have to compare to the existing table schema
+	hauserSchema, err := bigquery.InferSchema(bundleEvent{})
+	if err != nil {
+		return err
+	}
+
+	if bq.doesTableExist(bq.conf.BigQuery.ExportTable) {
+		log.Printf("Export table exists, making sure the schema in BigQuery is compatible with the schema specified in Hauser")
+
+		// get current table schema in BigQuery
+		table := bq.bqClient.Dataset(bq.conf.BigQuery.Dataset).Table(bq.conf.BigQuery.ExportTable)
+		md, err := table.Metadata(context.Background())
+		if err != nil {
+			return err
+		}
+
+		// Find the fields from the hauser schema that are missing from the BiqQuery table
+		missingFields := bq.GetMissingFields(hauserSchema, md.Schema)
+
+		// If fields are missing, we add them to the table schema
+		if len(missingFields) > 0 {
+			// Append missing fields to export table schema
+			md.Schema = bq.AppendToSchema(md.Schema, missingFields)
+			if err := bq.updateTable(table, md.Schema); err != nil {
+				return nil
+			}
+		}
+	} else {
+		// Table does not exist, create new table
+		log.Printf("Export table does not exist, creating one.")
+		if err := bq.createExportTable(hauserSchema); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (bq *BigQuery) updateTable (table *bigquery.Table, schema bigquery.Schema) error {
+	// update the table so we update the schema
+	log.Printf("Updating Export table schema")
+	tmd := bigquery.TableMetadataToUpdate {Schema: schema}
+	if _, err := table.Update(bq.ctx, tmd, ""); err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetMissingFields returns all fields that are present in the hauserSchema, but not in the bqSchema
+func (bq *BigQuery) GetMissingFields(hauserSchema, bqSchema bigquery.Schema) []*bigquery.FieldSchema {
+	bqSchemaMap := makeSchemaMap(bqSchema)
+	var missingFields []*bigquery.FieldSchema
+	for _, f := range hauserSchema {
+		if _, ok := bqSchemaMap[f.Name]; !ok {
```

Comment:
If I'm reading [this doc](https://cloud.google.com/bigquery/docs/schemas#column_names) correctly, BQ column names are also case-insensitive. Interestingly, BQ does appear to consistently keep whatever case was used when creating the table, but I think we still need to handle a possible case difference.

Reply:
good catch! Updated the column comparisons to be case insensitive!